r1cs/gadgets/std/src/bits/macros.rs

```diff
@@ -0,0 +1,2532 @@
+                    } else {
+                        rhs
+                    };
+
+                    let bits = vec![Boolean::constant(false); by]
+                        .iter() // append rhs zeros as least significant bits
+                        .chain(self.bits.iter()) // Chain existing bits as most significant bits starting from least significant ones
+                        .take($bit_size) // Crop after $bit_size bits
+                        .map(|el| *el)
+                        .collect();
+
+                    Self {
+                        bits,
+                        value: self.value.map(|v| v << by as $native_type),
+                    }
+                }
+            }
+
+            impl Shr<usize> for $type_name {
+                type Output = Self;
+
+                fn shr(self, rhs: usize) -> Self::Output {
+                    let by = if rhs >= $bit_size {
+                        $bit_size-1
+                    } else {
+                        rhs
+                    };
+
+                    let bits = self
+                    .bits
+                    .iter()
+                    .skip(by) // skip least significant bits which are removed by the shift
+                    .chain(vec![Boolean::constant(false); by].iter()) // append zeros as most significant bits
+                    .map(|el| *el)
+                    .collect();
+
+                    Self {
+                        bits,
+                        value: self.value.map(|v| v >> by as $native_type),
+                    }
+                }
+            }
+
+
+            impl RotateUInt for $type_name {
+                fn rotl(&self, by: usize) -> Self {
+                    let by = by % $bit_size;
+
+                    let bits = self
+                    .bits
+                    .iter()
+                    .skip($bit_size - by)
+                    .chain(self.bits.iter())
+                    .take($bit_size)
+                    .map(|el| *el)
+                    .collect();
+
+                    Self {
+                        bits,
+                        value: self.value.map(|v| v.rotate_left(by as u32)),
+                    }
+                }
+
+                fn rotr(&self, by: usize) -> Self {
+                    let by = by % $bit_size;
+
+                    let bits = self
+                    .bits
+                    .iter()
+                    .skip(by)
+                    .chain(self.bits.iter())
+                    .take($bit_size)
+                    .map(|el| *el)
+                    .collect();
+
+                    Self {
+                        bits,
+                        value: self.value.map(|v| v.rotate_right(by as u32)),
+                    }
+                }
+            }
+
+            //this macro allows to implement the binary bitwise operations already available for Booleans (i.e., XOR, OR, AND)
+            macro_rules! impl_binary_bitwise_operation {
```

Comment:
I tried it but it is a mess with trait bounds on the function for the gadget op, because I need to call the gadget function with a namespace rather than a constraint system. If not strictly necessary, I would keep the macro